src/filters/common.js

```diff
@@ -0,0 +1,99 @@
+/**
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module paste-from-office/filters/common
+ */
+
+/**
+ * Removes paragraph wrapping content inside list item.
+ *
+ * @param {module:engine/view/element~Element|module:engine/view/documentfragment~DocumentFragment} elementOrDocumentFragment
+ * @param {module:engine/view/upcastwriter~UpcastWriter} writer
+ */
+export function unwrapParagraph( elementOrDocumentFragment, writer ) {
+	const iterableNodes = elementOrDocumentFragment.is( 'element' ) ? elementOrDocumentFragment.getChildren() : elementOrDocumentFragment;
+
+	for ( const element of iterableNodes ) {
+		if ( element.is( 'element', 'p' ) && element.parent.is( 'element', 'li' ) ) {
+			unwrapSingleElement( element, writer );
+		}
+
+		if ( element.is( 'element' ) && element.childCount ) {
+			unwrapParagraph( element, writer );
```

Comment:
Similarly to the method below, I think that we can get rid of recursion here by inspectign children of the `li` rahter then parent of `p`.